django/db/models/functions.py

```diff
@@ -26,8 +26,10 @@ class ToNCLOB(Func):
 
             expressions = [
                 ToNCLOB(expression) for expression in self.get_source_expressions()]
-            self.set_source_expressions(expressions)
-        return super(Coalesce, self).as_sql(compiler, connection)
+            clone = self.copy()
```

Comment:
To test you'd need a model with a CharField and a TextField which only runs on Oracle. Then you'd have to evaluate the queryset, and check the expression for `ToNCLOB` which is a private class. Otherwise you'd require two different backends to be active within the test suite and do something like:

```
qs = Article.objects.annotate(Coalesce('char_field', 'text_field))
oracle = list(qs.using('oracle'))
postgres = list(qs.using('postgres'))  # would fail without the clone 
```

In the longer term I think it'd be nice if we could do mixed backend testing (where `other` is a sqlite backend for example). That's probably too ambitious for this PR though.